File: tests/party-log.py
#!/usr/bin/env python3
"""The party's notes as a log — what the vault stores, and what the pages show.

Two halves, both run from here:

  * the vault half, in this file: the markdown a party note is stored as, and
    what happens to its timestamp when a player rewrites their own line. This
    is where the risk is — the file is edited in Obsidian too, so anything
    written has to parse back the way it went in.
  * the page half, in tests/party-log.mjs: the same modules the browser loads,
    in a jsdom document, fed by the fixture server started here.

    python3 tests/party-log.py

Exits non-zero on the first failure. tests/session.py remains the one that
proves any of this against a real canvas.
"""
import json
import os
import re
import shutil
import subprocess
import sys
import tempfile
import threading
import time
import http.server
import urllib.error
import urllib.parse
import urllib.request

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
sys.path.insert(0, ROOT)
import notes_api as N                                          # noqa: E402

PORT = 8913
BASE = f'http://127.0.0.1:{PORT}'
MAP = '/maps/Realm.png'

fails = 0


def check(ok, what, detail=''):
    global fails
    if not ok:
        fails += 1
    print(('  ok   ' if ok else '  BAD  ') + what
          + (('\n         ' + str(detail)) if detail and not ok else ''))


# A map with a history: two lines from before timestamps existed, a session's
# worth of dated ones, and one that was edited the next morning.
FIXTURE = '''---
map: "/maps/Realm.png"
---
# Party Notes — Realm.png

## 1,1
- **Old Hand** — written before times were kept

## 3,4
- **Dame Ada** (2026-09-16 20:10) — a ruined tower
- **Sir Tim** (2026-09-16 21:05 · edited 2026-09-17 00:50) — we camped by the ford

## 9,9
- **Knyan** (2026-09-16 20:42) — hoofprints, many
'''


class Handler(http.server.BaseHTTPRequestHandler):
    def is_remote(self):
        return False

    def do_POST(self):
        p = urllib.parse.urlparse(self.path)
        if not N.handle_post(self, p):
            self.send_response(404); self.end_headers()

    def do_GET(self):
        p = urllib.parse.urlparse(self.path)
        if not N.handle_get(self, p):
            self.send_response(404); self.end_headers()

    def log_message(self, *a):
        pass


def post(cell, text, by, at=None, expect=200):
    body = {'map': MAP, 'cell': cell, 'text': text, 'by': by}
    if at is not None:
        body['at'] = at
    req = urllib.request.Request(
        BASE + '/api/partynotes', data=json.dumps(body).encode(),
        headers={'Content-Type': 'application/json'})
    try:
        return 200, json.loads(urllib.request.urlopen(req).read())
    except urllib.error.HTTPError as e:
        return e.code, json.loads(e.read() or b'{}')


def cells():
    return json.loads(urllib.request.urlopen(
        BASE + '/api/partynotes?map=' + urllib.parse.quote(MAP)).read())['cells']


def main():
    vault = tempfile.mkdtemp(prefix='gmd-party-log-')
    N._vault_root = vault
    os.makedirs(os.path.join(vault, N.SUBDIR), exist_ok=True)
    path = N.party_path(MAP)
    with open(path, 'w', encoding='utf-8') as f:
        f.write(FIXTURE)

    srv = http.server.ThreadingHTTPServer(('127.0.0.1', PORT), Handler)
    threading.Thread(target=srv.serve_forever, daemon=True).start()
    time.sleep(0.3)

    try:
        print('\n1. The markdown round-trips')
        # Everything the parser can be handed, including the shapes a GM
        # typing into Obsidian will produce.
        sample = {
            '1,1': [{'by': 'Ada', 'text': 'plain, undated'}],
            '2,2': [{'by': 'Tim', 'text': 'dated', 'at': '2026-09-16 20:10'}],
            '3,3': [{'by': 'Knyan', 'text': 'edited', 'at': '2026-09-16 20:10',
                     'edited': '2026-09-17 09:00'}],
            '4,4': [{'by': 'Ada', 'text': 'the bridge — what is left of it — sags',
                     'at': '2026-09-16 20:11'}],
            '5,5': [{'by': 'Ada', 'text': '(we think) a cairn'}],
        }
        back = N._parse_party(N._compose_party(MAP, sample))
        check(back == sample, 'compose -> parse gives back exactly what went in', back)

        print('\n2. Writing a note records when it was written')
        post('7,7', 'a standing stone', 'Dame Ada')
        e = cells()['7,7'][0]
        check(e.get('at', '').startswith(time.strftime('%Y-%m-%d %H:%M', time.localtime())),
              'in local wall-clock', e)

        print('\n3. A hex accumulates — a second note does not replace the first')
        # This is the whole point: you come back to a hex having seen something
        # new. The old observation is still true.
        post('7,7', 'and a second stone, fallen', 'Dame Ada')
        got = cells()['7,7']
        check(len(got) == 2, 'the same player can write twice on one hex', got)
        check([x['text'] for x in got] ==
              ['a standing stone', 'and a second stone, fallen'],
              'oldest first, both intact', got)
        check(got[0]['at'] != got[1]['at'],
              'and the two are individually addressable', got)

        post('7,7', 'I saw it too', 'Knyan')
        check(len(cells()['7,7']) == 3, 'as can somebody else', cells()['7,7'])

        print('\n4. Your own entry is still yours to change or take back')
        at0 = cells()['7,7'][0]['at']
        post('7,7', 'a standing stone, mossy', 'Dame Ada', at=at0)
        got = cells()['7,7']
        check(got[0]['text'] == 'a standing stone, mossy', 'editing changes that one', got[0])
        check(got[0]['at'] == at0, 'and it keeps its place in the log', got[0])
        check(got[0].get('edited'), 'while recording that it changed', got[0])
        check(len(got) == 3, 'without disturbing the others', got)

        # An entry is addressed by WHO wrote it as well as when, so one player
        # cannot reach another's — not even when the two were written in the
        # same second and share a timestamp.
        knyan = next(x for x in cells()['7,7'] if x['by'] == 'Knyan')
        st, _ = post('7,7', 'not mine to touch', 'Dame Ada', at='1999-01-01 00:00:00')
        check(st == 404, 'editing an entry that does not exist is refused', st)
        post('7,7', 'not mine to touch', 'Mallory', at=knyan['at'])
        still = next(x for x in cells()['7,7'] if x['by'] == 'Knyan')
        check(still['text'] == knyan['text'],
              "and another player's entry is left alone", still)

        post('7,7', '', 'Dame Ada', at=at0)
        got = cells()['7,7']
        check(len(got) == 2 and all(x['text'] != 'a standing stone, mossy' for x in got),
              'and an empty text removes that one entry', got)

        print('\n5. An empty note is nothing to say, not a delete')
        st, d = post('7,7', '   ', 'Knyan')
        check(st == 400, 'adding blank text is refused', (st, d))
        check(len(cells()['7,7']) == 2, 'and removes nothing', cells()['7,7'])

        print('\n6. The vault file is still something you would read in Obsidian')
        with open(path, encoding='utf-8') as f:
            lines = [l for l in f.read().split('\n') if l.startswith('- **')]
        for l in lines[:4]:
            print('         ' + l)
        check(all(re.match(r'^- \*\*[^*]+\*\*( \([^)]*\))? — ', l) for l in lines),
              'every line is name, when, then the note', lines)

        # --- the page half ---
        print('\n--- the pages (tests/party-log.mjs) ---')
        env = dict(os.environ, PARTY_LOG_BASE=BASE)
        r = subprocess.run(['node', os.path.join(HERE, 'party-log.mjs')],
                           cwd=HERE, env=env)
        if r.returncode:
            print('\nthe page half failed')
            return 1
    finally:
        srv.shutdown()
        shutil.rmtree(vault, ignore_errors=True)

    print(f'\nvault failures: {fails}')
    return 1 if fails else 0


if __name__ == '__main__':
    sys.exit(main())

File: tests/party-log.mjs
// The party log, on both pages, driven through the real modules in a real DOM.
//
// Not a browser — jsdom — so this asserts wiring and rendering rather than
// pixels: that the button toggles the panel, that the rows come out in the
// order the notes were written, that a player's text cannot inject markup,
// and that a row takes you back to its cell. tests/session.py is still the
// one that proves it against a real canvas.
//
//     python3 tests/party-log.py        # starts the fixture server, runs this
import { JSDOM } from 'jsdom';
import { readFileSync } from 'fs';

const BASE = process.env.PARTY_LOG_BASE || 'http://127.0.0.1:8913';
const MAP = '/maps/Realm.png';

let fail = 0;
const check = (ok, what, detail = '') => {
  if (!ok) fail++;
  console.log((ok ? '  ok   ' : '  BAD  ') + what + (!ok && detail ? '\n         ' + detail : ''));
};

// The real player page, minus its module script (we import the modules here),
// plus the one GM control this file also exercises.
const remoteHtml = readFileSync(new URL('../remote.html', import.meta.url), 'utf8')
  .replace(/<script[\s\S]*?<\/script>/g, '');
const dom = new JSDOM(remoteHtml.replace('</body>',
  '<button id="btn-party-log"></button>'
  + '<div id="party-log" class="party-log" style="display:none;"></div></body>'),
  { url: BASE });

// jsdom ships no canvas. The pips are drawn on one and this file is not about
// pixels, so hand out a context that accepts every call and remembers nothing.
// Without it the party-notes listener throws before it reaches the panel.
const NOOP_CTX = new Proxy({}, {
  get: (t, k) => (k in t ? t[k]
    : (t[k] = typeof k === 'string' && /^(canvas)$/.test(k) ? null : () => NOOP_CTX)),
  set: () => true,
});
dom.window.HTMLCanvasElement.prototype.getContext = () => NOOP_CTX;

globalThis.window = dom.window;
globalThis.document = dom.window.document;
globalThis.requestAnimationFrame = (fn) => setTimeout(fn, 0);
globalThis.cancelAnimationFrame = clearTimeout;
globalThis.localStorage = dom.window.localStorage;
const realFetch = globalThis.fetch;
globalThis.fetch = (u, o) => realFetch(String(u).startsWith('http') ? u : BASE + u, o);

const PN = await import('../js/party-notes.js');
const { S } = await import('../js/store.js');

PN.setPartyNotesMap(MAP);
await new Promise(r => setTimeout(r, 250));
await PN.loadPartyNotes(true);

console.log('\n1. The order is the order they were written');
const log = PN.partyLog();
log.forEach(e => console.log(`     ${(e.at || '(undated)').padEnd(17)} ${e.cell.padEnd(5)} ${e.by}: ${e.text}`));
const dated = log.filter(e => e.at).map(e => e.at);
check(dated.length > 1, 'the fixture has something to order', String(dated.length));
check(dated.every((v, i) => i === 0 || dated[i - 1] <= v),
      'entries run oldest to newest across cells', JSON.stringify(dated));
check(log.findIndex(e => e.at) > log.map(e => !e.at).lastIndexOf(true),
      'undated entries come first — they predate anything dated');
const seen = new Set(log.map(e => e.cell + '|' + e.by));
check(seen.size === log.length, 'no entry appears twice');

console.log('\n2. What a row shows');
const html = PN.partyLogHTML();
check(/pl-time">\d{2}:\d{2}</.test(html), 'the time it was written');
check(html.includes('data-cell='), 'the cell, as a link back to the map');
check(html.includes('pl-day'), 'a date heading over each day');
check(html.includes('Before times were kept'),
      'and undated entries say so rather than borrowing a time');
check(html.includes('· edited'), 'an entry changed later is marked edited');

console.log('\n3. A note is player-typed text on a shared screen');
await PN.savePartyNote('5,5', '<img src=x onerror=alert(1)> & "q"', 'Knyan');
await PN.loadPartyNotes(true);
const evil = PN.partyLogHTML();
check(!evil.includes('<img src=x'), 'markup in a note is escaped');
check(evil.includes('&lt;img'), 'and shown as text');
const box = document.getElementById('party-log');
box.innerHTML = evil;
check(box.querySelectorAll('img').length === 0, 'and creates no element when parsed');

console.log('\n4. The GM button toggles the panel');
const CN = await import('../js/cell-notes.js');
const B = await import('../js/bindings.js');
S.lastMapSrc = MAP;
// Both pages' bindings, so a button wired into the wrong one is caught.
B.attachGmBindings();
B.attachRemoteBindings();
const unbound = B.unboundSelectors().join(' ');
check(!unbound.includes('#btn-party-log'), 'the GM button is actually bound', unbound);
check(!unbound.includes('#remote-log-btn'), 'and so is the player button', unbound);
const btn = document.getElementById('btn-party-log');
const click = (id) => document.getElementById(id).dispatchEvent(
  new dom.window.MouseEvent('click', { bubbles: true }));
click('btn-party-log');
check(box.style.display === 'block', 'first click opens it', box.style.display);
check(box.querySelectorAll('.pl-row').length === log.length + 1,
      'with a row per note', String(box.querySelectorAll('.pl-row').length));
check(/Hide party log/.test(btn.textContent), 'and the button says how to close it',
      btn.textContent);
click('btn-party-log');
check(box.style.display === 'none', 'second click closes it', box.style.display);
check(/Party log/.test(btn.textContent) && /\d/.test(btn.textContent),
      'and it goes back to naming the count', btn.textContent);

console.log('\n5. The player button toggles its sheet, and only one sheet at a time');
const RP = await import('../js/remote-page.js');
const sheet = document.getElementById('remote-log-panel');
click('remote-log-btn');
check(sheet.style.display === 'block', 'the log opens', sheet.style.display);
check(document.getElementById('remote-log-btn').classList.contains('active'),
      'and the button reads as on');
click('remote-legend-btn');
check(sheet.style.display === 'none', 'opening the legend closes the log', sheet.style.display);
click('remote-log-btn');
check(document.getElementById('remote-legend-panel').style.display === 'none',
      'and opening the log closes the legend');
RP.remoteToggleNotes();
check(sheet.style.display === 'none', 'as does opening notes', sheet.style.display);

console.log('\n6. Each name wears that player\'s colour');
const tinted = PN.partyLogHTML((who) => who === 'Knyan' ? '#22c55e' : '');
check(tinted.includes('style="color:#22c55e;"'), 'the author is tinted');
check((tinted.match(/style="color:/g) || []).length === 
      (PN.partyLog().filter(e => e.by === 'Knyan').length),
      'and only the players that have one are');

console.log('\n7. A hex accumulates notes');
// The player sheet's job changed: the box adds a note, it does not rewrite
// the one you left there last time. Drive it the way a finger does.
S.myActorName = 'Knyan';
S.remoteFrame = { width: 1000, height: 800 };
const CELL = '12,12';
await PN.addPartyNote(CELL, 'first thing I saw', 'Knyan');
await PN.addPartyNote(CELL, 'and later, a second', 'Knyan');
await PN.loadPartyNotes(true);
check(PN.partyNotesFor(CELL).length === 2, 'two notes from one player coexist',
      JSON.stringify(PN.partyNotesFor(CELL)));

const rendered = PN.partyNotesFor(CELL);
check(rendered[0].text === 'first thing I saw' && rendered[1].text === 'and later, a second',
      'and they stay in the order they were written', JSON.stringify(rendered));

// Editing one leaves the other alone.
await PN.editPartyNote(CELL, rendered[0].at, 'first thing I saw, corrected', 'Knyan');
await PN.loadPartyNotes(true);
const after = PN.partyNotesFor(CELL);
check(after.length === 2, 'editing one does not remove the other', JSON.stringify(after));
check(after[0].text === 'first thing I saw, corrected', 'the right one changed', JSON.stringify(after[0]));
check(after[1].text === 'and later, a second', 'the other is untouched', JSON.stringify(after[1]));

// Removing one leaves the other alone.
await PN.removePartyNote(CELL, after[0].at, 'Knyan');
await PN.loadPartyNotes(true);
const left = PN.partyNotesFor(CELL);
check(left.length === 1 && left[0].text === 'and later, a second',
      'and removing one leaves the rest', JSON.stringify(left));

// An empty add is nothing to say, and must not be read as a delete.
const before = PN.partyNotesFor(CELL).length;
await PN.addPartyNote(CELL, '   ', 'Knyan');
await PN.loadPartyNotes(true);
check(PN.partyNotesFor(CELL).length === before,
      'an empty note deletes nothing', String(PN.partyNotesFor(CELL).length));

console.log('\n7b. The player sheet lists what is on the hex, and adds to it');
// Tap a hex for real: notes mode on, then a click on the note canvas. jsdom
// reports a zero-origin rect, so with scale 1 and no crop, client coords are
// map coords and the cell a tap lands in is predictable.
S.remoteScale = 1;
S.remoteCrop = { x: 0, y: 0 };
S.remoteFit = { w: 1000, h: 800 };
S.tokenGridEnabled = true;
S.tokenGridShow = true;
S.keyCellPx = 100; S.keyCellYPx = 0; S.keyOx = 0; S.keyOy = 0;
S.tokenGridType = 'square';
S.myActorName = 'Knyan';
RP.setupRemoteView();
if (!document.getElementById('remote-note-panel').innerHTML) RP.remoteToggleNotes();

const canvas = document.getElementById('remote-note-canvas');
canvas.dispatchEvent(new dom.window.MouseEvent('click',
  { bubbles: true, clientX: 250, clientY: 250 }));
await new Promise(r => setTimeout(r, 30));
const panel = document.getElementById('remote-note-panel');
const tapped = (/Cell ([-\d,]+)/.exec(panel.textContent) || [])[1];
check(!!tapped, 'tapping a hex binds the sheet to it', panel.textContent.slice(0, 80));

if (tapped) {
  await PN.addPartyNote(tapped, 'a ford, shallow', 'Knyan');
  await PN.addPartyNote(tapped, 'and a heron', 'Knyan');
  await PN.addPartyNote(tapped, 'I saw the heron too', 'Dame Ada');
  await PN.loadPartyNotes(true);
  await new Promise(r => setTimeout(r, 30));

  const p = document.getElementById('remote-note-panel');
  check(p.querySelectorAll('.rn-entry').length === 3,
        'the sheet lists every note on the hex, not just yours',
        String(p.querySelectorAll('.rn-entry').length));
  check(p.querySelectorAll('.rn-entry.mine').length === 2,
        'marking which are yours', String(p.querySelectorAll('.rn-entry.mine').length));
  check(p.querySelectorAll('[data-edit]').length === 2
        && p.querySelectorAll('[data-del]').length === 2,
        'with edit and remove on yours only');
  check(/Add note/.test(p.textContent),
        'and the box adds a new one rather than rewriting the last',
        p.textContent.slice(-120));
  const ta = document.getElementById('remote-note-text');
  check(ta.value === '', 'so it starts empty', JSON.stringify(ta.value));

  // Adding through the button, the way a player does.
  ta.value = 'the heron left';
  ta.dispatchEvent(new dom.window.Event('input', { bubbles: true }));
  document.getElementById('remote-note-save').dispatchEvent(
    new dom.window.MouseEvent('click', { bubbles: true }));
  await new Promise(r => setTimeout(r, 200));
  check(PN.partyNotesFor(tapped).length === 4,
        'the button adds a fourth note', JSON.stringify(PN.partyNotesFor(tapped).map(e => e.text)));
  check(PN.partyNotesFor(tapped).filter(e => e.by === 'Knyan').length === 3,
        'all three of yours still there', JSON.stringify(PN.partyNotesFor(tapped)));

  // Editing one of yours through its button.
  const p2 = document.getElementById('remote-note-panel');
  p2.querySelector('[data-edit]').dispatchEvent(
    new dom.window.MouseEvent('click', { bubbles: true }));
  await new Promise(r => setTimeout(r, 30));
  const ta2 = document.getElementById('remote-note-text');
  check(ta2.value === 'a ford, shallow', 'the edit button loads that note into the box',
        JSON.stringify(ta2.value));
  check(/Save change/.test(document.getElementById('remote-note-panel').textContent),
        'and the button changes what it will do');
  ta2.value = 'a ford, shallow and stony';
  ta2.dispatchEvent(new dom.window.Event('input', { bubbles: true }));
  document.getElementById('remote-note-save').dispatchEvent(
    new dom.window.MouseEvent('click', { bubbles: true }));
  await new Promise(r => setTimeout(r, 200));
  const texts = PN.partyNotesFor(tapped).map(e => e.text);
  check(texts.includes('a ford, shallow and stony') && texts.length === 4,
        'saving changes that one and only that one', JSON.stringify(texts));

  // Removing one of yours.
  const p3 = document.getElementById('remote-note-panel');
  p3.querySelector('[data-del]').dispatchEvent(
    new dom.window.MouseEvent('click', { bubbles: true }));
  await new Promise(r => setTimeout(r, 200));
  check(PN.partyNotesFor(tapped).length === 3,
        'the remove button takes back one note',
        JSON.stringify(PN.partyNotesFor(tapped).map(e => e.text)));
  check(PN.partyNotesFor(tapped).some(e => e.by === 'Dame Ada'),
        "and never somebody else's");
}

console.log('\n8. Both of a player\'s notes reach the log');
const log2 = PN.partyLog().filter(e => e.cell === CELL);
check(log2.length === 1, 'the log carries what is left', JSON.stringify(log2));
await PN.addPartyNote(CELL, 'a third, much later', 'Knyan');
await PN.loadPartyNotes(true);
const log3 = PN.partyLog().filter(e => e.cell === CELL);
check(log3.length === 2, 'and grows as notes are added', JSON.stringify(log3));
check(log3[0].at <= log3[1].at, 'still oldest first', JSON.stringify(log3.map(e => e.at)));

console.log('\nFAILURES: ' + fail);
process.exit(fail ? 1 : 0);
